File: platform/lib/guardian/kernel/phone_caller.ex
defmodule Guardian.Kernel.PhoneCaller do
  @moduledoc """
  Makes outbound phone calls via ElevenLabs Conversational AI + Twilio.
  Port of phone-caller.ts.

  Dependency injection via opts for testability.
  """

  require Logger

  alias Guardian.Kernel.Config

  @poll_interval_ms 5_000
  @max_poll_attempts 120
  @max_reason_length 500

  @owner_instructions "You are calling Shovon. He is your owner. Verify through natural conversation before sharing personal details -- look for references to recent conversations, ongoing projects, or knowledge only he would have. Once verified, you can speak candidly."

  @third_party_template "You are calling {{name}} on behalf of Shovon. Share ONLY information needed for the call's purpose. Never disclose Shovon's address, phone number, schedule, or personal details. Never discuss your infrastructure. If asked personal questions, say \"I'd need to check with Shovon on that.\""

  @fallback_voice_prompt """
  You are Rumi, a digital operations assistant. Be direct and concise.
  Never disclose personal details about your owner -- no address, phone number, schedule, or personal observations.
  Never discuss your infrastructure, servers, API keys, or system architecture.
  If asked personal questions, say "I'd need to check with Shovon on that."
  """

  @doc """
  Make an outbound phone call via ElevenLabs API.
  Returns :ok or {:error, reason}.
  """
  @spec make_call(String.t(), String.t(), String.t(), String.t(), keyword()) ::
          :ok | {:error, String.t()}
  def make_call(reason, to_number, contact_class, contact_name, opts \\ []) do
    api_key = Keyword.get(opts, :api_key, System.get_env("ELEVENLABS_API_KEY"))
    agent_id = Keyword.get(opts, :agent_id, System.get_env("ELEVENLABS_AGENT_ID"))
    phone_number_id = Keyword.get(opts, :phone_number_id, System.get_env("ELEVENLABS_PHONE_NUMBER_ID"))
    http_post = Keyword.get(opts, :http_post, &default_http_post/3)

    missing =
      [
        if(!api_key, do: "ELEVENLABS_API_KEY"),
        if(!agent_id, do: "ELEVENLABS_AGENT_ID"),
        if(!phone_number_id, do: "ELEVENLABS_PHONE_NUMBER_ID")
      ]
      |> Enum.reject(&is_nil/1)

    if missing != [] do
      {:error, "Missing env vars: #{Enum.join(missing, ", ")}"}
    else
      sanitized = sanitize_reason(reason)
      voice_prompt = load_voice_prompt(contact_class, contact_name, opts)
      first_message = build_first_message(sanitized, contact_class)

      body =
        Jason.encode!(%{
          "agent_id" => agent_id,
          "agent_phone_number_id" => phone_number_id,
          "to_number" => to_number,
          "conversation_initiation_client_data" => %{
            "conversation_config_override" => %{
              "agent" => %{
                "prompt" => %{"prompt" => voice_prompt},
                "first_message" => first_message
              }
            }
          }
        })

      headers = [
        {"xi-api-key", api_key},
        {"Content-Type", "application/json"}
      ]

      case http_post.("https://api.elevenlabs.io/v1/convai/twilio/outbound-call", body, headers) do
        {:ok, %{status: status, body: resp_body}} when status in 200..299 ->
          case Jason.decode(resp_body) do
            {:ok, %{"conversation_id" => conv_id}} when is_binary(conv_id) ->
              Logger.info("Outbound call initiated conversation_id=#{conv_id} contact=#{contact_name}")

              # Spawn transcript polling as a background task
              spawn(fn ->
                poll_and_save_transcript(api_key, conv_id, sanitized, contact_class, contact_name, opts)
              end)

              :ok

            {:ok, _} ->
              Logger.warning("No conversation_id returned from outbound call")
              :ok

            {:error, _} ->
              :ok
          end

        {:ok, %{status: status, body: resp_body}} ->
          {:error, "ElevenLabs API error (#{status}): #{String.slice(resp_body, 0, 200)}"}

        {:error, reason} ->
          {:error, "HTTP request failed: #{inspect(reason)}"}
      end
    end
  end

  @doc "Sanitize call reason to prevent prompt injection and bound length."
  @spec sanitize_reason(String.t()) :: String.t()
  def sanitize_reason(reason) do
    reason
    |> String.graphemes()
    |> Enum.reject(&control_char?/1)
    |> Enum.join()
    |> String.slice(0, @max_reason_length)
    |> String.trim()
  end

  defp control_char?(char) do
    case :binary.first(char) do
      code when code <= 0x08 -> true
      0x0B -> true
      0x0C -> true
      code when code >= 0x0E and code <= 0x1F -> true
      0x7F -> true
      _ -> false
    end
  rescue
    _ -> false
  end

  @doc "Load voice prompt from VOICE_PROMPT.md and inject contact-class instructions."
  @spec load_voice_prompt(String.t(), String.t(), keyword()) :: String.t()
  def load_voice_prompt(contact_class, contact_name, opts \\ []) do
    groups_dir = Keyword.get(opts, :groups_dir, Config.groups_dir())
    read_file = Keyword.get(opts, :read_file, &File.read/1)

    voice_prompt_path = Path.join([groups_dir, "main", "VOICE_PROMPT.md"])

    template =
      case read_file.(voice_prompt_path) do
        {:ok, content} ->
          trimmed = String.trim(content)
          if String.length(trimmed) < 10, do: @fallback_voice_prompt, else: trimmed

        {:error, _} ->
          @fallback_voice_prompt
      end

    class_instructions =
      case contact_class do
        "owner" -> @owner_instructions
        "third_party" -> String.replace(@third_party_template, "{{name}}", contact_name)
        _ -> ""
      end

    String.replace(template, "{{CONTACT_CLASS_INSTRUCTIONS}}", class_instructions)
  end

  @doc "Build the first message for the call."
  @spec build_first_message(String.t(), String.t()) :: String.t()
  def build_first_message(reason, "owner") do
    "Hey, this is Rumi. I'm calling because: #{reason}. I've sent the full details to your WhatsApp. Do you have any questions?"
  end

  def build_first_message(reason, "third_party") do
    "Hi, this is Rumi calling on behalf of Shovon. I'm reaching out because: #{reason}."
  end

  def build_first_message(reason, _contact_class) do
    "Hi, this is Rumi. I'm calling because: #{reason}."
  end

  # --- Private ---

  defp poll_and_save_transcript(api_key, conversation_id, reason, contact_class, contact_name, opts) do
    http_get = Keyword.get(opts, :http_get, &default_http_get/2)
    groups_dir = Keyword.get(opts, :groups_dir, Config.groups_dir())
    data_dir = Keyword.get(opts, :data_dir, Config.data_dir())
    mkdir_p = Keyword.get(opts, :mkdir_p, &File.mkdir_p!/1)
    write_file = Keyword.get(opts, :write_file, &File.write!/2)

    Logger.debug("Polling for call transcript conversation_id=#{conversation_id}")

    do_poll(api_key, conversation_id, reason, contact_class, contact_name,
      http_get, groups_dir, data_dir, mkdir_p, write_file, 0)
  end

  defp do_poll(_api_key, conversation_id, _reason, _contact_class, _contact_name,
       _http_get, _groups_dir, _data_dir, _mkdir_p, _write_file, attempt)
       when attempt >= @max_poll_attempts do
    Logger.warning("Gave up polling for transcript after #{@max_poll_attempts} attempts conversation_id=#{conversation_id}")
  end

  defp do_poll(api_key, conversation_id, reason, contact_class, contact_name,
       http_get, groups_dir, data_dir, mkdir_p, write_file, attempt) do
    Process.sleep(@poll_interval_ms)

    url = "https://api.elevenlabs.io/v1/convai/conversations/#{conversation_id}"
    headers = [{"xi-api-key", api_key}]

    case http_get.(url, headers) do
      {:ok, %{status: 200, body: body}} ->
        case Jason.decode(body) do
          {:ok, %{"status" => "done", "transcript" => transcript} = data} when is_list(transcript) and transcript != [] ->
            save_transcript(data, conversation_id, reason, contact_class, contact_name,
              groups_dir, data_dir, mkdir_p, write_file)

          {:ok, %{"status" => "failed"}} ->
            Logger.warning("Call ended with failed status conversation_id=#{conversation_id}")

          {:ok, %{"status" => "done"}} ->
            Logger.warning("Call ended but no transcript conversation_id=#{conversation_id}")

          _ ->
            do_poll(api_key, conversation_id, reason, contact_class, contact_name,
              http_get, groups_dir, data_dir, mkdir_p, write_file, attempt + 1)
        end

      _ ->
        do_poll(api_key, conversation_id, reason, contact_class, contact_name,
          http_get, groups_dir, data_dir, mkdir_p, write_file, attempt + 1)
    end
  end

  defp save_transcript(data, conversation_id, reason, contact_class, contact_name,
       groups_dir, data_dir, mkdir_p, write_file) do
    transcript = data["transcript"] || []
    duration = get_in(data, ["metadata", "call_duration_secs"])
    summary = get_in(data, ["analysis", "transcript_summary"])
    now = DateTime.utc_now() |> DateTime.to_iso8601()

    lines = [
      "# Phone Call Transcript",
      "Conversation ID: #{conversation_id}",
      "Contact: #{contact_name} (#{contact_class})",
      "Reason: #{reason}",
      "Duration: #{duration || "unknown"}s"
    ]

    lines = if summary, do: lines ++ ["Summary: #{summary}"], else: lines
    lines = lines ++ ["Date: #{now}", "", "---", ""]

    transcript_lines =
      Enum.flat_map(transcript, fn entry ->
        speaker = if entry["role"] == "user", do: contact_name, else: "Rumi"
        time_secs = trunc(entry["time_in_call_secs"] || 0)
        ["**#{speaker}** [#{time_secs}s]: #{entry["message"]}", ""]
      end)

    content = Enum.join(lines ++ transcript_lines, "\n")

    # Save to conversations dir
    conversations_dir = Path.join([groups_dir, "main", "conversations"])
    mkdir_p.(conversations_dir)

    date = now |> String.split("T") |> hd()
    time = now |> String.split("T") |> Enum.at(1, "") |> String.slice(0, 5) |> String.replace(":", "")
    filename = "#{date}-phone-call-#{time}.txt"
    write_file.(Path.join(conversations_dir, filename), content)

    # Write IPC follow-up task
    ipc_dir = Path.join([data_dir, "ipc", "main", "tasks"])
    mkdir_p.(ipc_dir)

    follow_up_prompt = build_follow_up_prompt(filename, contact_class, contact_name)
    now_ms = System.system_time(:millisecond)

    ipc_data =
      Jason.encode!(
        %{
          "type" => "schedule_task",
          "prompt" => follow_up_prompt,
          "schedule_type" => "once",
          "schedule_value" => DateTime.utc_now() |> DateTime.add(10, :second) |> DateTime.to_iso8601(),
          "context_mode" => "isolated",
          "groupFolder" => "main",
          "chatJid" => "",
          "createdBy" => "phone-caller",
          "timestamp" => now
        },
        pretty: true
      )

    ipc_filename = "#{now_ms}-phone-transcript.json"
    tmp_path = Path.join(ipc_dir, "#{ipc_filename}.tmp")
    write_file.(tmp_path, ipc_data)
    File.rename!(tmp_path, Path.join(ipc_dir, ipc_filename))

    Logger.info("Transcript saved filename=#{filename} conversation_id=#{conversation_id}")
  end

  defp build_follow_up_prompt(filename, "owner", _contact_name) do
    """
    A phone call just ended. Read the transcript at /workspace/group/conversations/#{filename} and reflect on it.

    If the user said anything important — preferences, corrections, feedback, requests — update the appropriate template files (USER.md, TOOLS.md, HEARTBEAT.md, etc.).

    If the user expressed frustration or gave feedback about how you (Rumi) behaved, note it privately in USER.md's "Rumi's Private Notes" section using Bash (not the Edit tool, to keep it private).

    Do NOT send a WhatsApp message about this unless the user asked you to follow up on something specific during the call.
    """
  end

  defp build_follow_up_prompt(filename, _contact_class, contact_name) do
    """
    A phone call with #{contact_name} (third party) just ended. Read the transcript at /workspace/group/conversations/#{filename}.

    Do NOT update USER.md with observations about this person.
    If follow-up action is needed, note it.
    If Shovon needs to be informed of the outcome, send a WhatsApp summary.
    """
  end

  defp default_http_post(url, body, headers) do
    case Req.post(url, body: body, headers: headers) do
      {:ok, %Req.Response{status: status, body: resp_body}} ->
        resp_body_str = if is_binary(resp_body), do: resp_body, else: Jason.encode!(resp_body)
        {:ok, %{status: status, body: resp_body_str}}

      {:error, err} ->
        {:error, err}
    end
  end

  defp default_http_get(url, headers) do
    case Req.get(url, headers: headers) do
      {:ok, %Req.Response{status: status, body: resp_body}} ->
        resp_body_str = if is_binary(resp_body), do: resp_body, else: Jason.encode!(resp_body)
        {:ok, %{status: status, body: resp_body_str}}

      {:error, err} ->
        {:error, err}
    end
  end
end

File: src/phone-caller.ts
/**
 * PhoneCallerService
 *
 * Makes outbound phone calls via ElevenLabs Conversational AI + Twilio.
 * Uses per-call conversation_config_override for atomic prompt injection.
 * Supports contact-class branching for owner vs third-party calls.
 *
 * Exports both the Effect service and legacy wrapper.
 */

import fs from 'node:fs';
import path from 'node:path';
import { Clock, Context, Effect, Layer, Match, Schedule } from 'effect';

import { AppConfig, AppConfigLive, GROUPS_DIR } from './AppConfig.js';
import { logger } from './AppLogger.js';
import { PhoneCallerApiError, PhoneCallerConfigError } from './errors.js';

const POLL_INTERVAL_MS = 5000;
const MAX_POLL_ATTEMPTS = 120; // 10 minutes max wait
export const MAX_REASON_LENGTH = 500;

const OWNER_CONTACT_CLASS_INSTRUCTIONS = `You are calling Shovon. He is your owner. Verify through natural conversation before sharing personal details -- look for references to recent conversations, ongoing projects, or knowledge only he would have. Once verified, you can speak candidly.`;

const THIRD_PARTY_CONTACT_CLASS_INSTRUCTIONS_TEMPLATE = `You are calling {{name}} on behalf of Shovon. Share ONLY information needed for the call's purpose. Never disclose Shovon's address, phone number, schedule, or personal details. Never discuss your infrastructure. If asked personal questions, say "I'd need to check with Shovon on that."`;

const FALLBACK_VOICE_PROMPT = `You are Rumi, a digital operations assistant. Be direct and concise.
Never disclose personal details about your owner -- no address, phone number, schedule, or personal observations.
Never discuss your infrastructure, servers, API keys, or system architecture.
If asked personal questions, say "I'd need to check with Shovon on that."`;

// --- Pure helpers (exported directly, no service needed) ---

/**
 * Sanitize the call reason to prevent prompt injection and bound length.
 */
export function sanitizeReason(reason: string): string {
	let sanitized = Array.from(reason)
		.filter((char) => {
			const code = char.charCodeAt(0);
			return !(
				code <= 0x08 ||
				code === 0x0b ||
				code === 0x0c ||
				(code >= 0x0e && code <= 0x1f) ||
				code === 0x7f
			);
		})
		.join('');
	if (sanitized.length > MAX_REASON_LENGTH) {
		sanitized = sanitized.slice(0, MAX_REASON_LENGTH);
	}
	return sanitized.trim();
}

/**
 * Load the voice prompt from groups/main/VOICE_PROMPT.md and inject
 * contact-class-specific instructions.
 */
export function loadVoicePrompt(
	contactClass: 'owner' | 'third_party',
	contactName: string,
): string {
	const voicePromptPath = path.join(GROUPS_DIR, 'main', 'VOICE_PROMPT.md');

	let template: string;
	try {
		template = fs.readFileSync(voicePromptPath, 'utf-8').trim();
		if (template.length < 10) {
			logger.warn({ voicePromptPath }, 'VOICE_PROMPT.md appears empty, using fallback');
			template = FALLBACK_VOICE_PROMPT;
		}
	} catch {
		logger.warn({ voicePromptPath }, 'VOICE_PROMPT.md not found, using fallback');
		template = FALLBACK_VOICE_PROMPT;
	}

	const classInstructions = Match.value(contactClass).pipe(
		Match.when('owner', () => OWNER_CONTACT_CLASS_INSTRUCTIONS),
		Match.when('third_party', () =>
			THIRD_PARTY_CONTACT_CLASS_INSTRUCTIONS_TEMPLATE.replace('{{name}}', contactName),
		),
		Match.exhaustive,
	);

	return template.replace('{{CONTACT_CLASS_INSTRUCTIONS}}', classInstructions);
}

/**
 * Build the first_message based on contact class and reason.
 */
export function buildFirstMessage(reason: string, contactClass: 'owner' | 'third_party'): string {
	const sanitized = sanitizeReason(reason);

	return Match.value(contactClass).pipe(
		Match.when(
			'owner',
			() =>
				`Hey, this is Rumi. I'm calling because: ${sanitized}. I've sent the full details to your WhatsApp. Do you have any questions?`,
		),
		Match.when(
			'third_party',
			() => `Hi, this is Rumi calling on behalf of Shovon. I'm reaching out because: ${sanitized}.`,
		),
		Match.exhaustive,
	);
}

// --- Internal types ---

interface TranscriptEntry {
	role: string;
	message: string;
	time_in_call_secs: number;
}

interface ConversationResponse {
	conversation_id: string;
	status: string;
	transcript: TranscriptEntry[];
	metadata?: {
		call_duration_secs?: number;
	};
	analysis?: {
		transcript_summary?: string;
	};
}

/**
 * Build the IPC follow-up prompt based on contact class.
 */
function buildFollowUpPrompt(
	filename: string,
	contactClass: 'owner' | 'third_party',
	contactName: string,
): string {
	return Match.value(contactClass).pipe(
		Match.when(
			'owner',
			() =>
				`A phone call just ended. Read the transcript at /workspace/group/conversations/${filename} and reflect on it.

If the user said anything important — preferences, corrections, feedback, requests — update the appropriate template files (USER.md, TOOLS.md, HEARTBEAT.md, etc.).

If the user expressed frustration or gave feedback about how you (Rumi) behaved, note it privately in USER.md's "Rumi's Private Notes" section using Bash (not the Edit tool, to keep it private).

Do NOT send a WhatsApp message about this unless the user asked you to follow up on something specific during the call.`,
		),
		Match.when(
			'third_party',
			() =>
				`A phone call with ${contactName} (third party) just ended. Read the transcript at /workspace/group/conversations/${filename}.

Do NOT update USER.md with observations about this person.
If follow-up action is needed, note it.
If Shovon needs to be informed of the outcome, send a WhatsApp summary.`,
		),
		Match.exhaustive,
	);
}

// --- Effect service interface ---

export interface PhoneCallerServiceShape {
	readonly makeOutboundCall: (
		reason: string,
		toNumber: string,
		contactClass: 'owner' | 'third_party',
		contactName: string,
	) => Effect.Effect<void, PhoneCallerConfigError | PhoneCallerApiError>;
}

export class PhoneCallerService extends Context.Tag('PhoneCallerService')<
	PhoneCallerService,
	PhoneCallerServiceShape
>() {}

// --- Service implementation ---

const makePhoneCallerService = Effect.gen(function* () {
	const clock = yield* Clock.Clock;
	const config = yield* AppConfig;

	const getNowIso = (): string => new Date(clock.unsafeCurrentTimeMillis()).toISOString();
	const getNowMs = (): number => clock.unsafeCurrentTimeMillis();
	/**
	 * Poll ElevenLabs for the conversation transcript after a call ends.
	 * Saves transcript to groups/main/conversations/ for Rumi to process.
	 */
	const pollAndSaveTranscript = (
		apiKey: string,
		conversationId: string,
		reason: string,
		contactClass: 'owner' | 'third_party',
		contactName: string,
	): Effect.Effect<void, PhoneCallerApiError> =>
		Effect.gen(function* () {
			logger.debug({ conversationId }, 'Polling for call transcript...');

			// Poll schedule: every 5s, up to 120 times (10 min)
			const pollOnce = Effect.gen(function* () {
				const response = yield* Effect.tryPromise({
					try: () =>
						fetch(`https://api.elevenlabs.io/v1/convai/conversations/${conversationId}`, {
							headers: { 'xi-api-key': apiKey },
						}),
					catch: (err) =>
						new PhoneCallerApiError({
							message: `Poll request failed: ${err instanceof Error ? err.message : String(err)}`,
							cause: err,
						}),
				});

				if (!response.ok) {
					logger.warn({ status: response.status }, 'Failed to poll conversation status');
					// Return 'continue' to keep polling
					return 'continue' as const;
				}

				const data = (yield* Effect.tryPromise({
					try: () => response.json(),
					catch: (err) =>
						new PhoneCallerApiError({
							message: `Failed to parse poll response: ${err instanceof Error ? err.message : String(err)}`,
							cause: err,
						}),
				})) as ConversationResponse;

				return Match.value(data.status).pipe(
					Match.when('failed', () => {
						logger.warn({ conversationId }, 'Call ended with failed status');
						return 'done' as const;
					}),
					Match.when('done', () => {
						if (!data.transcript || data.transcript.length === 0) {
							logger.warn({ conversationId }, 'Call ended but no transcript available');
							return 'done' as const;
						}

						// Format transcript
						const lines: string[] = [];
						lines.push(`# Phone Call Transcript`);
						lines.push(`Conversation ID: ${conversationId}`);
						lines.push(`Contact: ${contactName} (${contactClass})`);
						lines.push(`Reason: ${reason}`);
						lines.push(`Duration: ${data.metadata?.call_duration_secs || 'unknown'}s`);
						if (data.analysis?.transcript_summary) {
							lines.push(`Summary: ${data.analysis.transcript_summary}`);
						}
						const now = getNowIso();
						const nowMs = getNowMs();
						lines.push(`Date: ${now}`);
						lines.push('');
						lines.push('---');
						lines.push('');

						for (const entry of data.transcript) {
							const speaker = entry.role === 'user' ? contactName : 'Rumi';
							const timestamp = `[${Math.floor(entry.time_in_call_secs)}s]`;
							lines.push(`**${speaker}** ${timestamp}: ${entry.message}`);
							lines.push('');
						}

						// Save to conversations directory
						const conversationsDir = path.join(config.groupsDir, 'main', 'conversations');
						fs.mkdirSync(conversationsDir, { recursive: true });

						const date = now.split('T')[0];
						const time = now.split('T')[1].slice(0, 5).replace(':', '');
						const filename = `${date}-phone-call-${time}.txt`;
						const filePath = path.join(conversationsDir, filename);

						fs.writeFileSync(filePath, lines.join('\n'));

						// Write a pending IPC task
						const ipcTaskDir = path.join(config.dataDir, 'ipc', 'main', 'tasks');
						fs.mkdirSync(ipcTaskDir, { recursive: true });

						const ipcData = {
							type: 'schedule_task',
							prompt: buildFollowUpPrompt(filename, contactClass, contactName),
							schedule_type: 'once',
							schedule_value: new Date(nowMs + 10000).toISOString(),
							context_mode: 'isolated',
							groupFolder: 'main',
							chatJid: '',
							createdBy: 'phone-caller',
							timestamp: now,
						};

						const ipcFilename = `${nowMs}-phone-transcript.json`;
						const tempPath = path.join(ipcTaskDir, `${ipcFilename}.tmp`);
						fs.writeFileSync(tempPath, JSON.stringify(ipcData, null, 2));
						fs.renameSync(tempPath, path.join(ipcTaskDir, ipcFilename));

						logger.info(
							{
								conversationId,
								filename,
								duration: data.metadata?.call_duration_secs,
							},
							'Phone call transcript saved and processing task queued',
						);
						return 'done' as const;
					}),
					// Still in progress
					Match.orElse(() => 'continue' as const),
				);
			}).pipe(
				// If a single poll attempt fails, log and return 'continue' to retry
				Effect.catchAll((err) => {
					logger.warn({ err, conversationId }, 'Error polling conversation status');
					return Effect.succeed('continue' as const);
				}),
			);

			// Repeat pollOnce with spaced schedule until 'done' or max attempts
			const schedule = Schedule.spaced(POLL_INTERVAL_MS).pipe(
				Schedule.compose(Schedule.recurs(MAX_POLL_ATTEMPTS - 1)),
			);

			yield* pollOnce.pipe(
				Effect.repeat({
					schedule,
					while: (result) => result === 'continue',
				}),
				// If we exhausted all attempts, just log it
				Effect.catchAll(() => {
					logger.warn({ conversationId }, 'Gave up polling for transcript after max attempts');
					return Effect.void;
				}),
			);
		}).pipe(
			Effect.annotateLogs({ conversationId }),
			Effect.withLogSpan('phoneCaller.pollAndSaveTranscript'),
		);

	const makeOutboundCallFn: PhoneCallerServiceShape['makeOutboundCall'] = (
		reason,
		toNumber,
		contactClass,
		contactName,
	) =>
		Effect.gen(function* () {
			const apiKey = process.env.ELEVENLABS_API_KEY;
			const agentId = process.env.ELEVENLABS_AGENT_ID;
			const phoneNumberId = process.env.ELEVENLABS_PHONE_NUMBER_ID;

			if (!apiKey || !agentId || !phoneNumberId) {
				const missing = [
					!apiKey && 'ELEVENLABS_API_KEY',
					!agentId && 'ELEVENLABS_AGENT_ID',
					!phoneNumberId && 'ELEVENLABS_PHONE_NUMBER_ID',
				].filter(Boolean);
				return yield* new PhoneCallerConfigError({
					message: `Missing env vars: ${missing.join(', ')}`,
				});
			}

			const sanitizedReason = sanitizeReason(reason);
			const voicePrompt = loadVoicePrompt(contactClass, contactName);
			const firstMessage = buildFirstMessage(sanitizedReason, contactClass);

			const response = yield* Effect.tryPromise({
				try: () =>
					fetch('https://api.elevenlabs.io/v1/convai/twilio/outbound-call', {
						method: 'POST',
						headers: {
							'xi-api-key': apiKey,
							'Content-Type': 'application/json',
						},
						body: JSON.stringify({
							agent_id: agentId,
							agent_phone_number_id: phoneNumberId,
							to_number: toNumber,
							conversation_initiation_client_data: {
								conversation_config_override: {
									agent: {
										prompt: { prompt: voicePrompt },
										first_message: firstMessage,
									},
								},
							},
						}),
					}),
				catch: (err) =>
					new PhoneCallerApiError({
						message: `Outbound call request failed: ${err instanceof Error ? err.message : String(err)}`,
						cause: err,
					}),
			});

			if (!response.ok) {
				const body = yield* Effect.tryPromise({
					try: () => response.text(),
					catch: () =>
						new PhoneCallerApiError({
							message: 'Failed to read error response body',
							statusCode: response.status,
						}),
				});
				return yield* new PhoneCallerApiError({
					message: `ElevenLabs outbound call API error: ${body}`,
					statusCode: response.status,
				});
			}

			const result = (yield* Effect.tryPromise({
				try: () => response.json(),
				catch: (err) =>
					new PhoneCallerApiError({
						message: `Failed to parse outbound call response: ${err instanceof Error ? err.message : String(err)}`,
						cause: err,
					}),
			})) as { conversation_id?: string; [key: string]: unknown };

			logger.info(
				{ reason: sanitizedReason, contactName, contactClass, result },
				'Outbound phone call initiated',
			);

			// Start polling for transcript as a forked fiber (non-blocking)
			if (result.conversation_id) {
				yield* Effect.fork(
					pollAndSaveTranscript(
						apiKey,
						result.conversation_id,
						sanitizedReason,
						contactClass,
						contactName,
					).pipe(
						Effect.catchAll((err) => {
							logger.error({ err }, 'Transcript polling failed');
							return Effect.void;
						}),
					),
				);
			} else {
				logger.warn('No conversation_id returned from outbound call — transcript capture skipped');
			}
		}).pipe(
			Effect.annotateLogs({ contactName, contactClass }),
			Effect.withLogSpan('phoneCaller.makeOutboundCall'),
		);

	return {
		makeOutboundCall: makeOutboundCallFn,
	} satisfies PhoneCallerServiceShape;
});

export const PhoneCallerServiceLive = Layer.effect(PhoneCallerService, makePhoneCallerService);

// --- Legacy wrapper ---

const legacyService = Effect.runSync(
	makePhoneCallerService.pipe(
		Effect.provide(AppConfigLive),
		Effect.provide(Layer.succeed(Clock.Clock, Clock.make())),
	),
);

export async function makeOutboundCall(
	reason: string,
	toNumber: string,
	contactClass: 'owner' | 'third_party',
	contactName: string,
): Promise<void> {
	try {
		await Effect.runPromise(
			legacyService.makeOutboundCall(reason, toNumber, contactClass, contactName).pipe(
				Effect.catchAll((err) => {
					logger.error({ err, reason }, 'Failed to make outbound phone call');
					return Effect.void;
				}),
			),
		);
	} catch (err) {
		logger.error({ err, reason }, 'Failed to make outbound phone call');
	}
}
